src/qml/qmlpreferencesproxy.cpp

```diff
@@ -0,0 +1,262 @@
+#include "qmlpreferencesproxy.h"
+
+#include <qqmlengine.h>
+
+#include <memory>
+
+#include "moc_qmlpreferencesproxy.cpp"
+#include "soundio/soundmanager.h"
+#include "soundio/soundmanagerutil.h"
+#include "util/assert.h"
+#include "util/scopedoverridecursor.h"
+
+namespace mixxx {
+namespace qml {
+
+namespace {
+const QString kAppGroup = QStringLiteral("[App]");
+const ConfigKey kKeylockEngineCfgkey =
+        ConfigKey(kAppGroup, QStringLiteral("keylock_engine"));
+
+} // namespace
+
+QmlSoundDeviceProxy::QmlSoundDeviceProxy(
+        SoundDevicePointer pInternal, Type type, QObject* parent)
+        : QObject(parent),
+          m_pInternal(pInternal),
+          m_type(type) {
+}
+uint QmlSoundDeviceProxy::getChannelCount() const {
+    return m_type == Type::Input ? m_pInternal->getNumInputChannels()
+                                 : m_pInternal->getNumOutputChannels();
+}
+SoundDeviceId QmlSoundDeviceProxy::getDeviceId() const {
+    return m_pInternal->getDeviceId();
+}
+
+QList<QmlSoundDeviceConnection*> QmlSoundDeviceProxy::connections(QmlSoundManagerProxy* manager) {
+    QList<QmlSoundDeviceConnection*> connections;
+
+    auto pManager = manager->internal();
+    auto config = pManager->getConfig();
+
+    if (m_type == Type::Input) {
+        const auto inputDeviceMap = config.getInputs();
+        for (auto it = inputDeviceMap.cbegin(); it != inputDeviceMap.cend(); ++it) {
+            if (it.key() == getDeviceId()) {
+                connections.push_back(new QmlSoundDeviceConnection(
+                        std::make_unique<AudioInput>(it.value()), this));
+            }
+        }
+    } else {
+        const auto ouputDeviceMap = config.getOutputs();
+        for (auto it = ouputDeviceMap.cbegin(); it != ouputDeviceMap.cend(); ++it) {
+            if (it.key() == getDeviceId()) {
+                connections.push_back(new QmlSoundDeviceConnection(
+                        std::make_unique<AudioOutput>(it.value()), this));
+            }
+        }
+    }
+    for (auto* pConnection : connections) {
+        QQmlEngine::setObjectOwnership(pConnection, QQmlEngine::JavaScriptOwnership);
+    }
+    return connections;
+}
+
+int QmlSoundDeviceConnection::getType() const {
+    return static_cast<int>(m_audioPath->getType());
+}
+uchar QmlSoundDeviceConnection::getChannelGroup() const {
+    auto group = m_audioPath->getChannelGroup();
+    return group.getChannelBase();
+}
+uchar QmlSoundDeviceConnection::getIndex() const {
+    return m_audioPath->getIndex();
+}
+
+QmlSoundManagerProxy::QmlSoundManagerProxy(
```

Comment:
I think it's better to have one file per object. Even though this will mean more, smaller files, it is easier to navigate the code and figure out where things are.